src/qibolab/_core/instruments/emulator/emulator.py

```diff
@@ -89,20 +89,20 @@ def _single_sequence(
         result for the execution of this single sequence, thus suitable
         to be returned as is.
         """
-        # probabilities for states in computational basis
-        probabilities = self._sweep(sequence, configs, sweepers)
-        assert options.nshots is not None
-        probabilities = np.moveaxis(probabilities, -3, 0)
+        # states in computational basis
+        states = self._sweep(sequence, configs, sweepers)
         probabilities = apply_to_last_two_axes(
             calculate_probabilities_density_matrix,
-            probabilities,
+            states,
             tuple(configs["hamiltonian"].single_qubit),
             configs["hamiltonian"].nqubits,
             configs["hamiltonian"].transmon_levels,
         )
-        averaged = shots(probabilities, options.nshots)
+
+        assert options.nshots is not None
+        sampled = shots(np.moveaxis(probabilities, -2, 0), options.nshots)
```

Comment:
I am not sure I follow this line. Based on my understanding of the code so far, this should simply be

```suggestion
        sampled = shots(probabilities, options.nshots)
```
since `probabilities` would already be in the correct format for `shots` function, which has shape (*batch_dimensions, d) where the density matrix is a d x d matrix.

Reply:
Explanation here: https://github.com/qiboteam/qibolab/pull/1209#discussion_r2069356826

You can do it before or after the `shots()` call, they commute. But you need to lift this dimension sooner or later (acquisitions are the innermost loop, since they are retrieved from a single evolution - but in Qibolab they are instead reported as the outermost dimension, even not as part of a multidimensional array, but structured as a `dict`, which is a one-dimensional sequence on its own).

However, I'm already trying to push these manipulations down as possible, to eventually avoid whatever we can. I'm just cautious, since confirming that everything is still working requires long runs. Thus, I was first focusing on preparing mock results to be able to confirm no disruption is happening (there is no intention of adding new features in this PR, but just making the code as little convoluted and efficient as possible).